### PR TITLE
Feature/313 - Add album details view

### PR DESCRIPTION
**Pull request by RomainNeup**
# Changelog
Doing the code to handle this feature request : https://github.com/Yooooomi/your_spotify/issues/313

## Backend Changes :gear:

- Added a new endpoint to retrieve album information:
  - The endpoint now returns album details, including the first and last track listening date, tracks, and the artists.

- Introduced a new endpoint to retrieve album rankings:
  - The endpoint now returns information on 3 albums: one ranked better, the current one, and the one immediately after it in the ranking.

- Updated the endpoint for retrieving artist statistics:
  - Added functionality to retrieve the Most Listened Album of the Artist.

## Frontend Changes :computer:

- Implemented a new album detail page:
  - Includes artist context along with tracks listened.
  - Displays album listened tracks.
  - Shows the first and last time each track was listened to.

- Added links to album pages wherever an album is mentioned across all pages.

- Added a most listened album panel to the artist page.

image

### CODE REVIEW

**Comment by RomainNeup** [comment]
Hello @Yooooomi could you have a look please :)

**Comment by quentinguidee** [comment]
cfr. https://github.com/Yooooomi/your_spotify/pull/348#issuecomment-1966614977

I think that all requests using `InfosModel` can use the performance boost from #348, however requests in this PR seem good for now since they are taking the same time than all other requests. I will probably try to refactor all requests at once after this PR is merged.

For the rest of this PR, the result seems pretty good! However I think there are some ESLint errors that need to be fixed before merging

File: client/src/components/InlineAlbum/InlineAlbum.tsx [hunk 1]
```diff
@@ -13,7 +15,9 @@ export default function InlineTrack<T extends HTMLTag = 'div'>({
   return (
     // eslint-disable-next-line react/jsx-props-no-spreading
     <Text {...other} title={album.name}>
-      {album.name}
+      <Link to={`/album/${album.id}`} className={s.root}>
+        {album.name}
+      </Link>
```
Comment (Yooooomi) [inline client/src/components/InlineAlbum/InlineAlbum.tsx:20]: Isn't that what InlineAlbum does?

File: client/src/components/Layout/Sider/Sider.tsx [hunk 2]
```diff
@@ -83,7 +91,11 @@ export default function Sider({ className, isDrawer }: SiderProps) {
       <div className={s.title}>
         <SiderTitle />
       </div>
-      <SiderSearch onArtistClick={goToArtist} onTrackClick={goToTrack} />
+      <SiderSearch
+        onArtistClick={goToArtist}
+        onTrackClick={goToTrack}
+        onAlbumClick={goToAlbum}
```
Comment (Yooooomi) [inline client/src/components/Layout/Sider/Sider.tsx:97]: Nitpick, but would be nicer to have onTrackClick, onAlbumClick, onArtistClick in this order, from small container to bigger

File: client/src/components/SiderSearch/SiderSearch.tsx [hunk 3]
```diff
@@ -12,12 +12,14 @@ import s from './index.module.css';
...
   onArtistClick?: (artist: Artist) => void;
   onTrackClick?: (track: TrackWithFullAlbum) => void;
+  onAlbumClick?: (album: Album) => void;
   inputClassname?: string;
 }
 
 export default function ArtistSearch({
   onArtistClick,
   onTrackClick,
+  onAlbumClick,
```
Comment (Yooooomi) [inline client/src/components/SiderSearch/SiderSearch.tsx:22]: same here

File: client/src/scenes/AlbumStats/AlbumStatsWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        </div>
+      </FullscreenCentered>
+    );
+  }
+
+  if ('code' in stats || !params.id) {
+    return (
+      <FullscreenCentered>
+        <Text element="h3">
+          You never listened to this song, might be someone else registered
```
Comment (Yooooomi) [inline client/src/scenes/AlbumStats/AlbumStatsWrapper.tsx:28]: You never listened to this **Album**

File: server/src/database/queries/artist.ts [hunk 5]
```diff
@@ -162,6 +162,38 @@ export const getTotalListeningOfArtist = async (
...
+  artistId: string,
+) => {
+  const res = await InfosModel.aggregate([
+    { $match: { owner: user._id } },
+    { $lookup: {
+      from: 'tracks',
+      localField: 'id',
+      foreignField: 'id',
+      as: 'track'
+    }},
```
Comment (Yooooomi) [inline server/src/database/queries/artist.ts:176]: yes indeed is the same pattern as #348

**Comment by Yooooomi** [comment]
I see you have the total time you listened to a specific album. Would be nice, for consistency, to have the same stat in the artist and track stat page (the times listened and total duration underneath). Nice design btw I like it :)

**Comment by Yooooomi** [comment]
Hello again. I realised I forgot to thank you for the job you've done. Thanks a lot man it means a lot that people get into the project to add new features!
I just pushed a new file architecture to the release 1.8.0 branch. You might want to merge this branch into yours since this merge request is going to target the 1.8.0 release branch (release/1.8.0).

**Comment by RomainNeup** [comment]
> I see you have the total time you listened to a specific album. Would be nice, for consistency, to have the same stat in the artist and track stat page (the times listened and total duration underneath). Nice design btw I like it :)

I will do it in another MR if you don't mind, as it has nothing to do with the topic of this one :)

**Comment by Yooooomi** [comment]
I kinda don't agree as consistency should be done as we develop new features so we don't pile up inconsistency debt.

But don't worry I'll take care of this beige releasing.

**Comment by RomainNeup** [comment]
> I kinda don't agree as consistency should be done as we develop new features so we don't pile up inconsistency debt.
> 
> But don't worry I'll take care of this before releasing.

I mean, I did it already but in another branch. So don't worry, I will just include it in this PR

**Comment by RomainNeup** [comment]
> cfr. [#348 (comment)](https://github.com/Yooooomi/your_spotify/pull/348#issuecomment-1966614977)
> 
> I think that all requests using `InfosModel` can use the performance boost from #348, however requests in this PR seem good for now since they are taking the same time than all other requests. I will probably try to refactor all requests at once after this PR is merged.
> 
> For the rest of this PR, the result seems pretty good! However I think there are some ESLint errors that need to be fixed before merging

Thanks 🙏 
For the ES lint errors, I don't have any on my side 🤔  Do you have a log you could provide or something ?

**Comment by quentinguidee** [comment]
I got them when starting the dev.sh containers. However that was before the recent changes in release/1.8.0, so maybe they are not present anymore

**Comment by Yooooomi** [comment]
Yeah eslint in 1.8.0 is a mess, I've set up a stricter minting than before so don't bother about it! Thanks a lot guys for the work. I'll try to merge them by tomorrow evening or so. You guys rock

**Comment by RomainNeup** [comment]
> Yeah eslint in 1.8.0 is a mess, I've set up a stricter linting than before so don't bother about it! Thanks a lot guys for the work. I'll try to merge them by tomorrow evening or so. You guys rock

I'm still trying to merge the branch into my branch. What a nightmare 🤯

File: client/src/components/SiderSearch/SiderSearch.tsx [hunk 6]
```diff
@@ -9,19 +9,19 @@ import Loader from '../Loader';
...
 
-export default function ArtistSearch({
-  onArtistClick,
+export default function SideSearch({
   onTrackClick,
+  onArtistClick,
   onAlbumClick,
   inputClassname,
-}: ArtistSearchProps) {
+}: SideSearchProps) {
```
Comment (Yooooomi) [inline client/src/components/SiderSearch/SiderSearch.tsx:24]: Typo, Sider

**Comment by Yooooomi** [comment]
> > Yeah eslint in 1.8.0 is a mess, I've set up a stricter linting than before so don't bother about it! Thanks a lot guys for the work. I'll try to merge them by tomorrow evening or so. You guys rock
> 
> I'm still trying to merge the branch into my branch. What a nightmare 🤯

I can take care of it if you want. Basically everything in client/ is now in apps/client

**Comment by RomainNeup** [comment]
Should be good but if you can check if I didn't break anything you made. Could be nice thanks 🙏

**Comment by RomainNeup** [comment]
Since the merge, I'm getting a `Error: No private data found, cannot sign JWT 2` do you know where it does come from ?

**Comment by Yooooomi** [comment]
> Since the merge, I'm getting a `Error: No private data found, cannot sign JWT 2` do you know where it does come from ?

It seems that you did not run migrations. Do you run the project using docker and dev.sh?

**Comment by Yooooomi** [comment]
I am also having issues with migrations. I'll keep in touch.

**Comment by Yooooomi** [comment]
I'm getting a blank page for album page, I'm trying to fix

**Comment by Yooooomi** [comment]
Everything works! Nice job, thanks for everything :)
Everything is ready for release 1.8.0. If you think we should wait for the time icon design before publishing it's ok otherwise we can go like this.

**Pull request merged**